rclcpp/src/rclcpp/context.cpp

```diff
@@ -341,20 +348,38 @@ Context::shutdown(const std::string & reason)
   return true;
 }
 
-rclcpp::Context::OnShutdownCallback
-Context::on_shutdown(OnShutdownCallback callback)
+rclcpp::OnShutdownCallbackHandle::SharedPtr
+Context::on_shutdown(OnShutdownCallbackHandle::OnShutdownCallbackType callback)
 {
-  on_shutdown_callbacks_.push_back(callback);
-  return callback;
+  std::lock_guard<std::mutex> lock(on_shutdown_callbacks_mutex_);
```

Comment:
Relatively minor, but this can be moved down to right above the `on_shutdown_callbacks_.push_back()`.  That is, we don't need to be holding this lock while we are allocating the handle and setting the callback.